Mathlib/CategoryTheory/CodiscreteCategory.lean

```diff
@@ -0,0 +1,161 @@
+  invFun := Codiscrete.mk
+  left_inv := by aesop_cat
+  right_inv := by aesop_cat
+
+instance {α : Type u} [DecidableEq α] : DecidableEq (Codiscrete α) :=
+  codiscreteEquiv.decidableEq
+
+namespace Codiscrete
+
+instance (A : Type*) : Category (Codiscrete A) where
+  Hom _ _ := Unit
+  id _ := ⟨⟩
+  comp _ _ := ⟨⟩
+
+section
+variable {C : Type u} [Category.{v} C] {A : Type w}
+
+/-- Any function `C → A` lifts to a functor `C ⥤  Codiscrete A`. For discrete categories this is
+called `functor` but we use that name for something else. -/
+def lift (F : C → A) : C ⥤ Codiscrete A where
+  obj := Codiscrete.mk ∘ F
+  map _ := ⟨⟩
+
+/-- Any functor `C ⥤  Codiscrete A` has an underlying function.-/
+def invlift (F : C ⥤ Codiscrete A) : C → A := Codiscrete.as ∘ F.obj
+
+/-- Given two functors to a codiscrete category, there is a trivial natural transformation.-/
+def natTrans {F G : C ⥤ Codiscrete A} :
+    F ⟶ G where
+  app _ := ⟨⟩
+
+/-- Given two functors into a codiscrete category, the trivial natural transformation is an
+natural isomorphism.-/
+def natIso {F G : C ⥤ Codiscrete A} :
+    F ≅ G where
+  hom := natTrans
+  inv := natTrans
+
+/-- Every functor `F` to a codiscrete category is naturally isomorphic {(actually, equal)} to
+  `Codiscrete.as ∘ F.obj`. -/
+@[simps!]
+def natIsoFunctor {F : C ⥤ Codiscrete A} : F ≅ lift (Codiscrete.as ∘ F.obj) := Iso.refl _
+
+end
+
+/-- A function induces a functor between codiscrete categories.-/
+def functorOfFun {A B : Type*} (f : A → B) : Codiscrete A ⥤ Codiscrete B := lift (f ∘ Codiscrete.as)
+
+open Opposite
+
+/-- A codiscrete category is equivalent to its opposite category. -/
+def oppositeEquivalence (A : Type*) : (Codiscrete A)ᵒᵖ ≌ Codiscrete A where
+  functor := lift (fun x ↦ Codiscrete.as x.unop)
+  inverse := (lift (fun x ↦ Codiscrete.as x.unop)).rightOp
+  unitIso := NatIso.ofComponents (fun _ => by exact Iso.refl _)
+  counitIso := natIso
+
+/-- Codiscrete.Functor turns a type into a codiscrete category-/
+def functor : Type u ⥤ Cat.{0,u} where
+  obj A := Cat.of (Codiscrete A)
+  map := functorOfFun
+
+open Adjunction Cat
+
+/-- For a category `C` and type `A`, there is an equivalence between functions `objects.obj C ⟶ A`
+and functors `C ⥤ Codiscrete A`.-/
+def equivFunctorToCodiscrete {C : Type u} [Category.{v} C] {A : Type w} :
+    (C → A) ≃ (C ⥤ Codiscrete A) where
+  toFun := lift
+  invFun := invlift
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+/-- The functor that turns a type into a codiscrete category is right adjoint to the objects
+functor.-/
+def adj : objects ⊣ functor := mkOfHomEquiv
+  {
+    homEquiv := fun _ _ => equivFunctorToCodiscrete
+    homEquiv_naturality_left_symm := fun _ _ => rfl
+    homEquiv_naturality_right := fun _ _ => rfl
+  }
+
+/--Unit of the adjunction Cat.objects ⊣ Codiscrete.functor -/
+def unit : 𝟭 Cat ⟶ objects ⋙ functor where
+  app := by
+    simp only [Functor.id_obj, Functor.comp_obj]
+    intro C
+    apply lift
+    exact fun a ↦ a
```

Comment:
We should not go into tactic mode when defining data. (There are only a few exceptions which are considered ok.)

For this constructions and other following similar patterns, I would suggest the following:
```lean
def unitApp (C : Type u) [Category.{v} C] : C ⥤ Codiscrete C := lift id

def counitApp (A : Type u) : Codiscrete A → A := Codiscrete.as

lemma adj_unit_app (X : Cat.{0, u}) :
    adj.unit.app X = unitApp X := rfl

lemma adj_counit_app (A : Type u) :
    adj.counit.app A = counitApp A := rfl
```
Then, for example `unitApp` is more universe polymorphic. I would argue that as much as possible, the definitions and lemmas should be done with the more general universes. Then, for defining stuff involving `Cat`, we specialized to the more constrained universe, as it is the case here because of the `0`.